src/vg/civcraft/mc/prisonpearl/managers/SummonManager.java

```diff
@@ -90,18 +144,91 @@ public boolean returnPlayer(PrisonPearl pearl) {
+		
+		Type t = null;
+		if (event == null) {
+			t = Type.RETURNED;
+		} else {
+			// Since there is a PlayerRespawnEvent we know that the player died and is being 
+			// returned that way.
+			t = Type.DIED;
+		}
+		
+		// Let's first deal with if the player is being returned to a different server.
+		Summon s = getSummon(pearl);
+		s.setTime(System.currentTimeMillis());
+		if (s.getReturnLocation() instanceof FakeLocation) {
+			// They are.
+			
+			if (pearled == null) {
+				// This happens when the command /ppreturn is used.
+				// The player may not be on this server but can now be returned.
+				// We want to send the request to all servers that the player should now be freed.
+				MercuryManager.returnPPSummon(pearl.getImprisonedId());
+				SummonEvent summonEvent = new SummonEvent(pearl, t);
+				Bukkit.getPluginManager().callEvent(summonEvent);
 				return true;
 			}
+			// Let's now check if they are even online.
+			PlayerDetails details = MercuryAPI.getServerforAccount(pearl.getImprisonedId());
+			
+			if (details != null) {
+				// They are online.
+				final FakeLocation loc = (FakeLocation) s.getReturnLocation();
+				storage.removeSummon(s);
+				SummonEvent summonEvent = new SummonEvent(pearl, t, pearled.getLocation());
+				Bukkit.getPluginManager().callEvent(summonEvent);
+				if (PrisonPearlConfig.getShouldPPReturnKill()) {
+					pearled.setHealth(0);
+				} else {
+					if (event != null) {
+						// This is being called from the player respawning.
+						Bukkit.getScheduler().runTask(PrisonPearlPlugin.getInstance(), new Runnable() {
+
+							@Override
+							public void run() {
+								try {
+									// Since the player is dead we actually want to random spawn them in that world.
+									BetterShardsAPI.randomSpawnPlayer(loc.getServerName(), pearled.getUniqueId());
+									BetterShardsAPI.connectPlayer(pearled, loc.getServerName(), PlayerChangeServerReason.PLUGIN);
+								} catch (PlayerStillDeadException e) {
+									// TODO Auto-generated catch block
+									e.printStackTrace();
+								}
+							}
+							
+						});
+						return true;
+					} else {
+						try {
+							TeleportInfo info = new TeleportInfo(loc.getWorldName(), loc.getServerName(), 
+									loc.getBlockX(), loc.getBlockY() + 1, loc.getBlockZ());
+							BetterShardsAPI.teleportPlayer(info.getServer(), pearled.getUniqueId(), info);
+							return BetterShardsAPI.connectPlayer(pearled, loc.getServerName(), PlayerChangeServerReason.PLUGIN);
+						} catch (PlayerStillDeadException e) {
+							// TODO Auto-generated catch block
+							e.printStackTrace();
+						}
+					}
+				}
+			}
+		} else {
+			SummonEvent summonEvent = new SummonEvent(pearl, t);
+			Bukkit.getPluginManager().callEvent(summonEvent);
+			// Player is being returned same server.
+			// Let's check if the player is online.
+			if (pearled == null) {
+				// They are not so lets just remove the summon.
+				storage.removeSummon(s);
```

Comment:
Just to check, this will result in them respawning back in the end when they next login?